csrc/scheduler/mma_utils.cpp

```diff
@@ -14,11 +15,37 @@
 #include <scheduler/utils.h>
 #include <variant>
 #include "mma_type.h"
-
 namespace nvfuser {
 
 namespace mma_utils {
 
+bool hasEnoughSharedMemoryForEpilogue(
+    const MatMulTileOptions& gemm_tile,
+    const int smem_double_buffer_stage) {
+  auto properties = at::cuda::getDeviceProperties(
+      c10::Device(c10::DeviceType::CUDA, 0).index());
+  const size_t device_smem_limit = properties->sharedMemPerBlockOptin;
+
+  // see scheduleContiguousVectorLoad
+  const int vector_word = 8;
+  auto warp_dims = gemm_tile.cta_tile / gemm_tile.warp_tile;
+  const int round_to_factor =
+      warp_dims.m * warp_dims.n * warp_dims.k * 32 * vector_word;
+  const int mk = gemm_tile.cta_tile.m * gemm_tile.cta_tile.k;
+  const int nk = gemm_tile.cta_tile.n * gemm_tile.cta_tile.k;
+  const size_t smem_a = (size_t)(ceilDiv(mk, round_to_factor) *
+                                 round_to_factor * smem_double_buffer_stage) *
+      dataTypeSize(DataType::Half);
+  const size_t smem_b = (size_t)(ceilDiv(nk, round_to_factor) *
```

Comment:
I'm concerned with hard-coded data types used for calculating the smem.

The information about `smem_a`, `smem_b` and `smem_c` data type can be access through this way, although it adds additional host overhead to calculate:

```
    const auto& roles_map_opt = mma_utils::getTensorsRoles(fusion);
    if (!roles_map_opt.isValid()) {
      return false; // some error or assert?
    }
    const auto& roles_map = roles_map_opt.getData();

    // Note, everything below can be enclosed in a lambda, where we capture roles, and pass desired role label and we get in return tensor type or it fails
    auto entry = roles_map.find(MatmulRole::INPUT_A);
    if (entry == toles_map.end()) {
        return false; // assert with message?
    }
    if (entry->second().empty()) {
       return false; // assert with message that there are no A inputs
    }

    const auto a_type = entry->second.front()->dtype();
```

To improve the execution flow / avoid accessing roles for heuristics, we could create roles map in the `getMatmulHeuristics(...)` and pass by reference this object here, to `hasEnoughSharedMemoryForEpilogue(...)`, this way roles will be available during whole heuristics process, if needed.

Reply:
accepted. To accommodate for the use in test cases, I add `std::vector<DataType> data_types` as an additional para to `hasEnoughSharedMemoryForEpilogue`